### PR TITLE
Support get environment variables in config

### PR DESCRIPTION
**Pull request by ysqi**
get environment variable if config item  has prefix <del>`$ENV_`   `$$env`</del> `${env}`. if environment variable is empty or not exist then return default value (split with `||`).
e.g.
```ini
token = ${TOKEN||astaxie}
[demo]
password = ${MyPWD}
user = ${MyUser||beego}
```
the result of  config.String("demo::password")  is the $MyPWD environment variable. 
some people need it , such as #1283 .

### CODE REVIEW

File: config/config.go [hunk 1]
```diff
@@ -107,6 +109,24 @@ func NewConfigData(adapterName string, data []byte) (Configer, error) {
...
+
+// Getenv return environment variable if env has prefix "$ENV_".
+func Getenv(env interface{}) (string, bool) {
+	if env == nil {
+		return "", false
+	}
+
+	// Onley support string key.
+	if key, ok := env.(string); ok {
+		if len(key) > len(envKeySign) && strings.HasPrefix(key, envKeySign) {
```
Comment (astaxie) [inline config/config.go]: you can remove `len(key) > len(envKeySign)`, because `strings.HasPrefix` already has that logic

File: config/config.go [hunk 2]
```diff
@@ -107,6 +109,24 @@ func NewConfigData(adapterName string, data []byte) (Configer, error) {
...
+			key = strings.TrimLeft(key, envKeySign)
```
Comment (astaxie) [inline config/config.go]: you should use TrimPrefix. TrimLeft is wrong

File: config/xml/xml.go [hunk 3]
```diff
@@ -190,7 +190,28 @@ func (c *ConfigContainer) DefaultStrings(key string, defaultval []string) []stri
 // GetSection returns map for the given section
 func (c *ConfigContainer) GetSection(section string) (map[string]string, error) {
 	if v, ok := c.data[section]; ok {
-		return v.(map[string]string), nil
+
+		var interfaceToStr func(map[string]interface{}) map[string]string
+
+		interfaceToStr = func(values map[string]interface{}) map[string]string {
```
Comment (fuxiaohei) [inline config/xml/xml.go]: write the function outside

**Comment by pjoe** [comment]
Just wondering the reason for not just using ExpandEnv as mentioned in #1283

Also for me it would be nice to be able to set a default (that is used if env var isn't set)

File: config/xml/xml.go (hunk 3 above)
Comment (ysqi) [inline config/xml/xml.go]: must?

File: config/config.go (hunk 2 above)
Comment (ysqi) [inline config/config.go]: yeh, you are reight.
```Go
// TrimPrefix returns s without the provided leading prefix string.
// If s doesn't start with prefix, s is returned unchanged.
func TrimPrefix(s, prefix string) string {
       if HasPrefix(s, prefix) {
             return s[len(prefix):]
       }
       return s
}
```

**Comment by ysqi** [comment]
@pjoe  The next step is to consider how to deal with.

**Comment by ysqi** [comment]
@pjoe  I think about it. Somtetime we can write some code to meet yourself demand. Not beego do but you do. You can DIY you config item.  such as:
**config**
```ini
defaultURL =  http://localhost
defaultPort = 8080
url = $ENV_SiteURL
port = $ENV_PORT
```

**Code**
```go
url := beego.AppConfig.DefaultString("url", "http://localhost")
if port := beego.AppConfig.String("port"); len(port) > 0 && port != "80" {
	url = fmt.Sprintf("%s:%s", url, port)
}
```
or 
```go
defaultPort := beego.AppConfig.String("defaultURL")
defaultURL := beego.AppConfig.String("defaultValue")

url := beego.AppConfig.DefaultString("url", defaultURL)
if port := beego.AppConfig.String("port"); len(port) > 0 && port != "80" {
	url = fmt.Sprintf("%s:%s", url, defaultPort)
}
```

File: config/xml/xml.go (hunk 3 above)
Comment (ysqi) [inline config/xml/xml.go]: @astaxie  What is your mind ?

**Comment by pjoe** [comment]
@ysqi: yeah i already have a 'solution' I'm using now, where I basically check env vars at startup and do beego.AppConfig.Set if env var exist. It feels a bit clumsy though.

However it does allow me to have different defaults for dev, staging, production (form config files) that are all overwritable by env vars.

File: config/xml/xml.go (hunk 3 above)
Comment (astaxie) [inline config/xml/xml.go]: I agree with @fuxiaohei. set the interfaceToStr as a function. that will be easily to understand. use too much closure. make the code unreadable

Comment (ysqi) [inline config/xml/xml.go]: OK , I will  remove interfaceToStr function to package config. I'm not sure it is a good idea.

**Comment by ysqi** [comment]
@pjoe  Maybe you can use different RunMode . such as :
```ini
httport = 6060
url = http://localhost
[dev]
httpport = 8080
[prod]
httpport = $ENV_PORT
url = $ENV_URL
[test]
httpport = 8888
```
you can call `beego.AppConfig.String("httpport")`  to get different default value for dev,test,production .

**Comment by pjoe** [comment]
@ysqi: I suppose that would work.

It does seem more natural for me to be able to set defaults, but maybe that can be added later as separate PR

**Comment by ysqi** [comment]
@astaxie  Pls review.

**Comment by miraclesu** [comment]
LGTM.

It will be great if the default value can be set likes:
```
httpport = $ENV_PORT|8080
url = $ENV_URL|http://www.default.com
```

**Comment by ysqi** [comment]
hi  @miraclesu , Unfortunately, it is not possible. but you can put it another way , such as like:
```ini 
[dev]
httpport = 8080
[prod]
httpport = $ENV_PORT
```

**Comment by JessonChan** [comment]
In fact,I'm not so sure about this.I would like to keep the config file just as simple as it should be.

**Comment by astaxie** [comment]
@ysqi please solve the conflict. and I think we need to discuss about the default value. That's useful. but we need to think about json,xml

File: config/config.go [hunk 4]
```diff
@@ -107,6 +107,63 @@ func NewConfigData(adapterName string, data []byte) (Configer, error) {
...
+//
+// Return environment variable if value start with "$$".
+// Return default value if environment variable is empty or not exist.
+//
+// It accept value formats "$$env" , "$$env||" , "$$env||defaultValue" , "defaultvalue".
+// Examples:
+//	v1 := config.ChooseRealValue("$$GOPATH")			// return the GOPATH environment variable.
+//	v2 := config.ChooseRealValue("$$GOAsta||/usr/local/go/")	// return the default value "/usr/local/go/".
+//	v3 := config.ChooseRealValue("Astaxie")				// return the value "Astaxie".
+func ChooseRealValue(value string) (realValue string) {
```
Comment (astaxie) [inline config/config.go]: I think we need to discuss this name. It not make sense for me. @pjoe do you have suggestion?

Comment (pjoe) [inline config/config.go]: Just at the top of my head: ExpandValueEnv or ConvertValueEnv, to make it sound similar to ExpandEnv builtin function (https://golang.org/pkg/os/#ExpandEnv)

**Comment by pjoe** [comment]
Just wondering how delimiters work with this. I would love to be able to write something like:
```
DbDsn=$$RDS_USERNAME:$$RDS_PASSWORD@tcp($$RDS_HOSTNAME:$$RDS_PORT)/$$RDS_DB_NAME
```

For e.g. setting database up for running on AWS Beanstalk with RDS

Maybe this should be delimited differently, like:
```
ConfigValue=${ENV_PART1||default1}/${ENV_PART2||default2}
```

Hope this makes sense :)

File: config/config.go (hunk 4 above)
Comment (ysqi) [inline config/config.go]: I like the function name "ExpandValueEnv"

**Comment by ysqi** [comment]
hey @pjoe  I can achieve next, First support use environment variables in config . a little every day :)

**Comment by astaxie** [comment]
@ysqi look through @pjoe idea. I think we maybe can change to `${ENV_PART1||default1}` for default value ENV. `${ENV_PART1}` for ENV.

If we want to support this style 

```
DbDsn=$$RDS_USERNAME:$$RDS_PASSWORD@tcp($$RDS_HOSTNAME:$$RDS_PORT)/$$RDS_DB_NAME
```

I think it's hard to implement this. As we don't know whether `:` is belong to ENV.

In future we can compatible with this config.
```
DbDsn=${RDS_USERNAME}:${RDS_PASSWORD}@tcp(${RDS_HOSTNAME}:${RDS_PORT})/${RDS_DB_NAME}
```

**Comment by pjoe** [comment]
@ysqi yeah better take small steps, but as @astaxie pointed out, having good delimiters around the env substitution is needed. Also I think `${ENV_VAR}` is a pretty common way of doing this

**Comment by ysqi** [comment]
@pjoe @astaxie Change key format use : ${ENV_PART||defaultValue} or  ${ENV_PART}.
Next time do more.

**Comment by astaxie** [comment]
@ysqi please also send a PR to update the documents

**Comment by pjoe** [comment]
@ysqi: congrats on finally getting this merged :D

Sorry for being a bit of a PITA, but wanted to make sure we got the foundation for this done right :)

**Comment by ysqi** [comment]
@pjoe  work for all for myself  :)

**Comment by osavchenko** [comment]
@ysqi are any documentation for this feature?

**Comment by ysqi** [comment]
@osavchenko ，you can find some in this commit include go  tes

and,look func comment at  config.go.
```go
// ExpandValueEnv returns value of convert with environment variable.
//
// Return environment variable if value start with "${" and end with "}".
// Return default value if environment variable is empty or not exist.
//
// It accept value formats "${env}" , "${env||}}" , "${env||defaultValue}" , "defaultvalue".
// Examples:
//	v1 := config.ExpandValueEnv("${GOPATH}")			// return the GOPATH environment variable.
//	v2 := config.ExpandValueEnv("${GOAsta||/usr/local/go}")	// return the default value "/usr/local/go/".
//	v3 := config.ExpandValueEnv("Astaxie")				// return the value "Astaxie".
```

**Comment by frederikhors** [comment]
@ysqi why nothing in docs about this? Is it still working?

**Comment by AndyZhaofeng** [comment]
My project has an .env file:
`
ADMIN_PORT=17201
ADMIN_ADDR=127.0.0.1
DATABASE_URL=127.0.0.1
MONGODB_URI=mongodb://127.0.0.1:27017
` 
and I want get values like this:
/conf/app.conf
`
AdminAddr = ${ADMIN_ADDR}
AdminPort = ${ADMIN_PORT}
`
when I execute the command : bee run
It throw an error, like this:
`
panic: prefix should has path
`
In my project it means, app.conf can not get .env vaules.
my beego version is beego/bee/v2 v2.1.0

**Comment by AndyZhaofeng** [comment]
> My project has an .env file:
> 
> ```
> ADMIN_PORT=17201
> 
> ADMIN_ADDR=127.0.0.1
> 
> DATABASE_URL=127.0.0.1
> 
> MONGODB_URI=mongodb://127.0.0.1:27017
> ```
> 
> and I want get values like this: /conf/app.conf
> 
> ```
> AdminAddr = ${ADMIN_ADDR}
> AdminPort = ${ADMIN_PORT}
> ```
> 
> when I execute the command : bee run It throw an error, like this:
> 
> ```
> panic: prefix should has path
> ```
> 
> In my project it means, app.conf can not get .env vaules. my beego version is beego/bee/v2 v2.1.0

import 
```
_ "github.com/joho/godotenv/autoload"
```
save that

**Pull request merged**